toolkit/tools/srpmpacker/srpmpacker.go

```diff
@@ -956,14 +958,30 @@ func hydrateFromRemoteSource(fileHydrationState map[string]bool, newSourceDir st
 
 		url := network.JoinURL(srcConfig.sourceURL, fileName)
 
-		err := retry.Run(func() error {
+		// Limit the number of concurrent network operations by pushing a struct{} into the channel. This will block until
+		// another operation completes and removes the struct{} from the channel.
+		if netOpsSemaphore != nil {
+			select {
+			case netOpsSemaphore <- struct{}{}:
+			case <-cancel:
+				logger.Log.Debug("Cancellation signal received at network operation semaphore")
+				return
+			}
+		}
+
+		err := retry.RunWithExpBackoff(func() error {
```

Comment:
Ideally the cancellation token should be checked during the retry back-off.

It seems like you can use the `time.After()` function in a `select` statement: https://golangbyexample.com/select-statement-with-timeout-go/

Reply:
So you are thinking an optional channel into the retry function? Maybe if its `nil` we make a temporary channel inside the function?
```go
func backoffSleep(delay time.Duration, cancelSignal <-chan struct{}) (cancelled bool)) {
	if cancelSignal == nil {
		cancelSignal = make(chan struct{})
	}
	select {
	case <-time.After(delay):
	case <-cancelSignal:
		cancelled = true
	}
	return
}

// runWithBackoffInternal runs function up to 'attempts' times, waiting delayCalc(failCount) before each i-th attempt.
// delayCalc(0) is expected to return 0.
func runWithBackoffInternal(function func() error, delayCalc func(failCount int) time.Duration, attempts int, cancelSignal <-chan struct{}) (err error) {
	for failures := 0; failures < attempts; failures++ {
		delayTime := delayCalc(failures)
		cancelled := backoffSleep(delayTime, cancelSignal)
		if cancelled {
			return fmt.Errorf("cancelled after %d attempts", failures)
		}
		if err = function(); err == nil {
			break
		}
	}
	return err
}
```